wrappers/python/c3py/c3py.py

```diff
@@ -667,16 +695,89 @@ def laplace_op(self, op, eps=0.0):
             ft_out.round(eps=eps)
 
         return ft_out
+    
+    def get_params(self):
+        assert self.ftp is not None
+        params = np.zeros(self.nparams)
+        c3.ft_param_get_params(self.ftp, params)
+        return params
+
+    def get_param(self, param_idx):
+        assert self.ftp is not None
+        return c3.ft_param_get_param(self.ftp, param_idx)
+        
+    def get_nparams(self):
+        assert self.ftp is not None
+        self.nparams = c3.ft_param_get_nparams(self.ftp)
+        return self.nparams
+
+    def update_params(self, new_params):
+        assert self.ftp is not None
+        c3.ft_param_update_params(self.ftp, new_params)
+        self.ft = c3.ft_param_get_ft(self.ftp)
+        
+    def param_grad_eval(self, x):
+        assert self.ftp is not None
+        grad = np.zeros(self.nparams)
+        c3.sl_mem_manager_check_structure(self.mem, self.ftp, x)
+        running_eval = c3.sl_mem_manager_get_running_eval(self.mem)
+        running_grad = c3.sl_mem_manager_get_running_grad(self.mem)
+        lin_structure_grad = c3.sl_mem_manager_get_lin_structure_vals(self.mem)
+        grad_eval = c3.ft_param_gradeval(self.ftp, x, grad, lin_structure_grad, running_grad, running_eval)
+        return grad_eval, grad
+    
+    def param_eval_lin(self, x):
+        assert self.ftp is not None
+        c3.sl_mem_manager_check_structure(self.mem, self.ftp, x)
+        running_eval = c3.sl_mem_manager_get_running_eval(self.mem)
+        lin_structure_grad = c3.sl_mem_manager_get_lin_structure_vals(self.mem)
+        lin_eval = c3.ft_param_eval_lin(self.ftp, lin_structure_grad, running_eval)
+        return lin_eval
+    
+    def param_grad_eval_lin(self, x):
+        assert self.ftp is not None
+        grad = np.zeros(self.nparams)
+        c3.sl_mem_manager_check_structure(self.mem, self.ftp, x)
+        running_eval = c3.sl_mem_manager_get_running_eval(self.mem)
+        running_grad = c3.sl_mem_manager_get_running_grad(self.mem)
+        lin_structure_grad = c3.sl_mem_manager_get_lin_structure_vals(self.mem)
+        grad_eval_lin = c3.ft_param_gradeval_lin(self.ftp, lin_structure_grad, grad, running_eval, running_grad)
+        return grad_eval_lin, grad
         
     def __del__(self):
         self.close()
 
     def close(self):
```

Comment:
Only FT should be deleted in the destructor. Everything else should be deleted after use within the functions. I do not see the need to add new deleters.